File: typescript/src/stack/valid-parentheses.ts
export function valid_parentheses(s: string): boolean {
  const stack: string[] = [];
  for (const c of s) {
    if (c === "(") {
      stack.push(")");
    } else if (c === "{") {
      stack.push("}");
    } else if (c === "[") {
      stack.push("]");
    } else if (stack.pop() !== c) {
      return false;
    }
  }
  return stack.length === 0;
}

File: rust/src/stack/valid_parentheses.rs
pub fn valid_parentheses(s: &str) -> bool {
    let mut stack = vec![];
    for b in s.bytes() {
        match b {
            b'(' => stack.push(b')'),
            b'{' => stack.push(b'}'),
            b'[' => stack.push(b']'),
            _ => {
                if stack.pop() != Some(b) {
                    return false;
                }
            }
        }
    }
    stack.is_empty()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_parentheses() {
        assert!(valid_parentheses("()"));
        assert!(valid_parentheses("()[]{}"));
        assert!(!valid_parentheses("(]"));
        assert!(valid_parentheses("([])"));
    }
}
